File: native/V2StyxLib/lib/utils/AbstractPoll.cpp
/*
 * AbstractPoll.cpp
 *
 */

#include "utils/AbstractPoll.h"

template <class T>
AbstractPoll<T>::AbstractPoll() {
}

template <class T>
AbstractPoll<T>::~AbstractPoll() {
}

template <class T>
T AbstractPoll<T>::getFreeItem() {

}

template <class T>
bool AbstractPoll<T>::release(T id) {

}

template <class T>
void AbstractPoll<T>::clean() {
	mAvailable.clear();
	mLast = 0;
}

File: java/v2styx-lib/src/main/java/com/v2soft/styxlib/utils/AbstractPoll.java
package com.v2soft.styxlib.utils;

import java.util.HashSet;

/**
 * @author V.Shcryabets <a>vshcryabets@gmail.com</a>
 */
public abstract class AbstractPoll<T> {
    protected HashSet<T> mAvailable = new HashSet<T>();
    protected T mLast = null;

    /**
     * @return Return free item
     */
    public T getFreeItem() {
        synchronized (mAvailable) {
            if (!mAvailable.isEmpty()) {
                T result = mAvailable.iterator().next();
                mAvailable.remove(result);
                return result;
            }
            mLast = getNext();
            return mLast;
        }
    }

    protected abstract T getNext();

    public boolean release(T id) {
        synchronized (mAvailable) {
            if ( mAvailable.contains(id)) {
                throw new IllegalStateException(
                        String.format("Something goes wrong, this item (%s) already has been released",
                                String.valueOf(id)));
            }
            return mAvailable.add(id);
        }
    }
    public void clean() {
        mAvailable.clear();
        mLast = null;
    }
}
